### ISSUE TITLE
Security issue with Docker

### ISSUE THREAD

**Issue by grossmj** [GNS3/gns3-server#1944]
It has been reported that it is really easy to access the host filesystem from inside a Docker container which obviously isn't something we want. Here is the report:

Just do: mount /dev/sda* /mnt
This will give you root access to the file system.

![docker_mount_test](https://user-images.githubusercontent.com/4752614/130399848-f96c445b-a00b-4b43-ad26-9ee3b0b6b2c1.png)

This isn’t a gns3-thing per-se and I don’t think rootless docker can do what you are doing with containers.

**Comment by b-ehlers** [comment]
The reason for this is the option `--privileged`, that GNS3 uses to start containers.

The same issue arises outside GNS3, when using `--privileged`, here an example:

```
behlers@iMac:~$ docker run -ti alpine
/ # mount /dev/sda4 /mnt
mount: permission denied (are you root?)
/ # exit
behlers@iMac:~$ docker run -ti --privileged alpine
/ # mount /dev/sda4 /mnt
/ # ls /mnt
bin             initrd.img.old  opt             sys
boot            lib             proc            tmp
dev             lib64           root            usr
etc             lost+found      run             var
home            media           sbin            vmlinuz
initrd.img      mnt             srv             vmlinuz.old
/ # umount /mnt
/ # exit
behlers@iMac:~$ 
```

To stop this, someone has to research, what capabilities are really needed for GNS3. Perhaps the handling of docker containers can be changed to run with fewer capabilities. Or some capabilities can be dropped after running the injected `/gns3/init.sh`. But all that is a lot of work.

**Comment by grossmj** [comment]
Thanks for the input. If I remember correctly, we need that to move interfaces to the network namespace of a container allowing it to communicate with other nodes in a GNS3 topology.

https://github.com/GNS3/gns3-server/blob/2.2/gns3server/compute/docker/docker_vm.py#L912L916

**Comment by b-ehlers** [comment]
I'm not sure. I think, that's done outside docker by ubridge. Therefore ubridge runs with the capabilities cap_net_admin and cap_net_raw.

I assume docker needs some privileges to set IP addresses and routes with the container. Also the handling of persistent directories (bind mount) may need some privileges. There may be more reasons for extended privileges, that needs some investigation.

**Comment by b-ehlers** [comment]
Checked some privilege needs of <https://github.com/GNS3/gns3-server/blob/master/gns3server/compute/docker/resources/init.sh>:

IP setup works by using only `cap_net_admin`:

```
behlers@iMac:~$ docker run -ti alpine
/ # ip link set eth0 down
ip: ioctl 0x8914 failed: Operation not permitted
/ # ip address add 1.1.1.1/24 dev eth0
ip: RTNETLINK answers: Operation not permitted
/ # ip route add 1.2.3.0/24 via 127.0.0.1
ip: RTNETLINK answers: Operation not permitted
/ # exit
behlers@iMac:~$ 
behlers@iMac:~$ docker run -ti --cap-add cap_net_admin alpine
/ # ip link set eth0 down
/ # ip link set eth0 up
/ # ip address add 1.1.1.1/24 dev eth0
/ # ip route add 1.2.3.0/24 via 127.0.0.1
/ # exit
behlers@iMac:~$ 
```

But most critical is the `mount --bind` in https://github.com/GNS3/gns3-server/blob/731152c75a2a6172b9b10ed381651fd812b87cbd/gns3server/compute/docker/resources/init.sh#L43

Allowing that mount would also allow any other mount, which results in this issue. So the implementation of persistent direcoties has to be rewritten to work without `mount --bind`.

**Comment by grossmj** [comment]
You are right, we need a workaround for `mount --bind "/gns3volumes$i" "$i"`

Also, I tried to at least run with without privileged flag by changing https://github.com/GNS3/gns3-server/blob/2.2/gns3server/compute/docker/docker_vm.py#L339L343 to this:

```
            "HostConfig": {
                "CapAdd": ["SYS_ADMIN", "NET_ADMIN"],
                "SecurityOptions": ["apparmor:unconfined"],  # https://github.com/moby/moby/issues/16429
                "Binds": mount_binds,
            },
```

Unfortunately, mount gets a permission deny. I tried a few workaround without any success...

![Screenshot from 2021-08-23 22-21-27](https://user-images.githubusercontent.com/4752614/130451380-93021135-776f-4d0c-bc2e-0893184ae569.png)

![Screenshot from 2021-08-23 22-20-59](https://user-images.githubusercontent.com/4752614/130451072-6cfe1ea3-f578-4023-93be-26e91cc366be.png)

**Comment by grossmj** [comment]
The only other solution I can think of would be to use volumes...

https://docs.docker.com/storage/volumes/

**Comment by grossmj** [comment]
This will have to wait for version >= 3.0 since there will be a lot of changes to make it work with volumes

**Comment by user32133** [comment]
Hello everyone, We discovered a way to access the host kernel using Docker escape by simply mounting the securityfs. 

The command is:
`mount -t securityfs securityfs /sys/kernel/security`

<img width="581" height="169" alt="Image" src="https://github.com/user-attachments/assets/b5a3e4ab-08ee-4119-b75b-f17f9d0457b0" />


It works on a clean Docker container of the `UbuntuDockerGuest`, installed from the GNS3.
[Reposted with identifying information redacted]

**Comment by user32133** [comment]
Hello everyone, the CVE-2026-30735 has been officially assigned for this issue.
@grossmj
